packages/core/micro/src/permutation/index.ts

```diff
@@ -0,0 +1,174 @@
+    order: 7,
+    emptySegment: empty,
+    deleteSegmentRange: (segment, start, length) => {
+        switch (segment.kind) {
+            case PermutationKind.Empty:
+                return emptyPair;
+
+            case PermutationKind.Direct:
+                return [
+                    segment,
+                    { kind: PermutationKind.Direct, content: segment.content.splice(start, length) }
+                ];
+
+            case PermutationKind.RunLength:
+                const content = segment.content;
+                const size = content.length;
+                let startIndex = 0;
+                let startPos = start;
+                while (startIndex < size) {
+                    const len = content[startIndex];
+                    if (startPos < len) {
+                        break;
+                    }
+                    startPos -= len;
+                    startIndex += 2;
+                }
+                let endIndex = 0;
+                let endPos = start + length;
+                while (endIndex < size) {
+                    const len = content[endIndex];
+                    if (endPos < len) {
+                        break;
+                    }
+                    endPos -= len;
+                    endIndex += 2;
+                }
+                if (startIndex === endIndex) {
+                    const len = content[startIndex];
+                    const base = content[startIndex + 1];
+                    const removed = content.splice(startIndex, 2, startPos, base, len - endPos, base);
+                    return [segment, { kind: PermutationKind.RunLength, content: removed }];
+                }
+                const wholeStartIndex = startPos === 0 ? startIndex : startIndex + 2;
+                const toRemove = (endIndex - wholeStartIndex) / 2;
+                endIndex -= toRemove;
+                const newContent: number[] = toRemove > 0 ? content.splice(wholeStartIndex, toRemove) : [];
+                if (startPos > 0) {
+                    newContent.push(content[startIndex] - startPos);
+                    newContent.push(content[startIndex + 1] + startPos);
+                    content[startIndex] = startPos;
+                }
+                if (endPos > 0) {
+                    newContent[newContent.length] = endPos;
+                    newContent[newContent.length] = content[endIndex + 1] - endPos;
+                    content[endIndex] -= endPos;
+                }
+                return [segment, { kind: PermutationKind.RunLength, content: newContent }];
+        }
+    }
+}
+
+export class IdentityVector {
+
+    permutationTree: AdjustTree<PermutationSegment> = createTree(permutationConfig);
+    next: number = 0;
+
+    getItem(position: number) {
+        const { offset, segment } = this.permutationTree.getItem(position);
+        switch (segment.kind) {
+            case PermutationKind.Empty:
+                return -1;
```

Comment:
I'm guessing 'empty' are a pair of initial placeholder segments that bookend the actual content?  That's a nice trick.

Reply:
I'm not sure it's doing anything that clever. It's just that the adjust tree is agnostic to the content (segments) at the leaves and when it needs to delete part of a leaf it calls deleteSegment, which should return the segment with the part removed, and the removed part. In the case of an empty segment which is unallocated / has no ID, I just use a singleton/sentinel.